src/fs.rs

```diff
@@ -0,0 +1,449 @@
+    /// Use [`LockingMode::Exclusive`] if you want to modify the file in any way.
+    /// Otherwise you should use [`LockingMode::Shared`].
+    /// `file_name` should just be a name, not a full path.
+    /// `file_flags` and `file_mode` are used for opening (and potentially creating) the file.
+    /// Note that this tries to create a separate lock file in the same directory as the file.
+    // TODO: set no-lock var when locking fails?
+    pub fn new(
+        locking_mode: LockingMode,
+        file_path: &wstr,
+        file_flags: OFlag,
+        file_mode: nix::sys::stat::Mode,
+    ) -> std::io::Result<Self> {
+        const LOCK_FILE_SUFFIX: &wstr = L!(".lock");
+        let mut lock_file_path = file_path.to_owned();
+        lock_file_path.push_utfstr(LOCK_FILE_SUFFIX);
+        let lock_file_path = lock_file_path;
+
+        if path_remoteness(&lock_file_path) == DirRemoteness::remote {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::Unsupported,
+                "Directory considered remote. Locking is disabled on remote file systems.",
+            ));
+        }
+
+        // We start by locking the dedicated lock file.
+        // This is required to avoid racing modifications by other threads/processes.
+        // If the lock is not released explicitly, it is released when the file descriptor of
+        // `lock_file` is closed, which happens in the `Drop` implementation of `File`.
+        let lock_file = wopen_cloexec(
+            &lock_file_path,
+            OFlag::O_RDONLY | OFlag::O_CREAT,
+            Mode::S_IRUSR,
+        )?;
+
+        // Try locking the file. Retry if locking was interrupted.
+        while unsafe { flock(lock_file.as_raw_fd(), locking_mode.to_flock_op()) } == -1 {
+            let err = std::io::Error::last_os_error();
+            if err.kind() != std::io::ErrorKind::Interrupted {
+                return Err(err);
+            }
+        }
+
+        // Open the data file
+        let data_file = wopen_cloexec(file_path, file_flags, file_mode)?;
+
+        Ok(Self {
+            data_file,
+            _lock_file: lock_file,
+        })
+    }
+
+    pub fn get(&self) -> &File {
+        &self.data_file
+    }
+
+    pub fn get_mut(&mut self) -> &mut File {
+        &mut self.data_file
+    }
+}
+
+/// Runs the `load` function, which should see a consistent state of the file at `path`.
+/// To ensure a consistent state, we use locks to prevent modifications by others.
+/// If locking is unavailable, the `load` function might be executed multiple times,
+/// until it manages a run without the file being modified in the meantime, or until the maximum
+/// number of allowed attempts is reached.
+/// If the file does not exist this function will return an error.
+pub fn lock_and_load<F, U>(
+    path: &wstr,
+    locking_enabled: &RelaxedAtomicBool,
+    load: F,
+) -> std::io::Result<(FileId, U)>
+where
+    F: Fn(&File) -> std::io::Result<U>,
```

Comment:
It's probably not immediately obvious that `U` is short for `Update` or `UpdateData`, so maybe use the long name.
It implies that it's used by the caller to update their in-memory state.. which is maybe not super obvious.
Alternative names might be `CapturedData`, `Extract`, ... but "update" is consistent with the rest of this patch.